src/controller/CHIPDeviceController.cpp

```diff
@@ -1639,18 +1639,46 @@ void DeviceCommissioner::CommissioningStageComplete(CHIP_ERROR err, Commissionin
 {
     // Once this stage is complete, reset mDeviceBeingCommissioned - this will be reset when the delegate calls the next step.
     MATTER_TRACE_EVENT_SCOPE("CommissioningStageComplete", "DeviceCommissioner");
-    if (mDeviceBeingCommissioned == nullptr)
+    NodeId nodeId       = kUndefinedNodeId;
+    DeviceProxy * proxy = nullptr;
+    if (mCommissioningDelegate != nullptr &&
+        mCommissioningDelegate->GetCommissioningParameters().GetNonConcurrentCommissioning().ValueOr(false))
     {
-        // We are getting a stray callback (e.g. due to un-cancellable
-        // operations) when we are not in fact commissioning anything.  Just
-        // ignore it.
-        return;
+        if (mDeviceBeingCommissioned != nullptr)
+        {
+            nodeId = mDeviceBeingCommissioned->GetDeviceId();
+            proxy  = mDeviceBeingCommissioned;
+            if (mCommissioningStage == kThreadNetworkEnable || mCommissioningStage == kWiFiNetworkEnable)
+            {
+                ReleaseCommissioneeDevice(reinterpret_cast<CommissioneeDeviceProxy *>(proxy));
```

Comment:
Both the `mDeviceBeingCommissioned` and `proxy` is type `DeviceProxy`. we should do the cast since the release function needs a `CommissioneeDeviceProxy` input.

Reply:
Why can't we make `mDeviceBeingCommissioned` a `CommissioneeDeviceProxy *`?  And if we can't, how do we know it's safe to cast it to that?

Reply:
Changed to release the device_poxy from the commissionee device pool.